// @cloudbase/ai/src/models/image-model.ts
import type {
  ModelReq,
  ImageGenerateInput,
  ImageEditInput,
  HunyuanGenerateImageInput,
  HunyuanGenerateImageOutput,
  HunyuanEditImageInput,
  HunyuanEditImageOutput,
} from '../type'

// mock NoInfer
type NoInfer<T> = [T][T extends any ? 0 : never]

type GenerateImageInput<TProvider extends string, TInput = unknown> = unknown extends TInput
  ? TProvider extends 'hunyuan-exp'
    ? HunyuanGenerateImageInput
    : ImageGenerateInput
  : TInput

type GenerateImageOutput<TProvider extends string, TOutput = unknown> = unknown extends TOutput
  ? TProvider extends 'hunyuan-exp'
    ? HunyuanGenerateImageOutput
    : unknown
  : TOutput

type EditImageInput<T extends string, TInput = unknown> = unknown extends TInput
  ? T extends 'hunyuan-exp'
    ? HunyuanEditImageInput
    : ImageEditInput
  : TInput

type EditImageOutput<T extends string, TOutput = unknown> = unknown extends TOutput
  ? T extends 'hunyuan-exp'
    ? HunyuanEditImageOutput
    : unknown
  : TOutput

export class DefaultImageModel<const TProvider extends string> {
  public generateImageSubUrl = 'images/generations'
  public editImageSubUrl = 'images/edits'

  constructor(
    private req: ModelReq,
    public baseUrl: string,
    public provider: TProvider,
  ) {}

  public async generateImage<TInput = unknown, TOutput = unknown>(input: NoInfer<GenerateImageInput<TProvider, TInput>>,): Promise<GenerateImageOutput<TProvider, TOutput>> {
    const res = await this.req({
      url: this.getGenerateUrl(),
      data: input,
      stream: false,
    })

    return res as GenerateImageOutput<TProvider, TOutput>
  }

  public async editImage<TInput = unknown, TOutput = unknown>(input: NoInfer<EditImageInput<TProvider, TInput>>,): Promise<EditImageOutput<TProvider, TOutput>> {
    const res = await this.req({
      url: this.getEditUrl(),
      data: input,
      stream: false,
    })
    return res as EditImageOutput<TProvider, TOutput>
  }

  private getGenerateUrl() {
    return `${this.baseUrl}/${this.provider}/${this.generateImageSubUrl}`
  }

  private getEditUrl() {
    return `${this.baseUrl}/${this.provider}/${this.editImageSubUrl}`
  }
}
